### ISSUE TITLE
Pre-processing of images before training/prediction

### ISSUE THREAD

**Issue by lizard24** [CSBDeep/CSBDeep#47]
Hello,

I am trying to figure out how the percentile normalisation of the images before training and prediction works - unfortunately, I find figuring this out from the source code is a bit too advanced for me.

Could you please explain to me the steps that are undertaken in default CARE? Example questions would be: Are the percentiles calculated for one image and then applied to each patch of this one image? Are similar percentiles then used for the prediction, or newly calculated for the specific image I want the network to predict?  I see that the functions in create_patches() use both, the input and ground truth images, so are the parameters for the normalisation of, let's say, the input images also dependent on the percentiles of the ground truth images?

Any comment would help!
Thank you very much,
Lisa

**Comment by maweigert** [comment]
Hi Lisa, 

You need to normalize the image by yourself before prediction, like so 

```python 
from csbdeep.utils import normalize 

img_normed = normalize(img, 1, 99.8)
```
By default, `create_patches` will randomly sample the min/max percentile from `(1,3)` and  `(99.5,99.9)` respectively,  so using above normalization at prediction time should be relatively safe (unless you suddenly predict on e.g. an almost empty image).

Hope that helps, 

Martin

**Comment by lizard24** [comment]
Hi Martin,

thanks for the quick reply!

Sorry, just to be clear - does that mean in default, CARE is not applying any form of pre-processing to the images before prediction?

Could I disable normalisation in create_patches ("normalization = no_norm") and use above normalisation function for the input/GT images before creating the rawdata object - and still get the same/very similar results?

Thanks a lot,
Lisa

**Comment by maweigert** [comment]
Ah, ok, I misunderstood your question.

> CARE is not applying any form of pre-processing to the images before prediction?

Contrary what I said before, the CARE models do in fact perform automatic percentile normalization (2,99.8) by default inside `model.predict`, so there is normally no need to do that explicitly. Is there any specific reason why you want to change that? 

> Could I disable normalisation in create_patches ("normalization = no_norm") and use above normalisation function for the input/GT images before creating the rawdata object - and still get the same/very similar results?

For the training to be successful the training input to the neural network has to be appropriately normalised. So the normalisation in `create_patches` should not be deactivated (unless of course your raw data is already normalised).

Hope that helps,
Martin

**Comment by lizard24** [comment]
Hi Martin,

thanks! The reason why I am asking: I would like to compare the performance of CARE with the performance of another network with different structure, since I am not explicitly using the CARE network for denoising - and I want to make sure it is still the best option. In order to make the comparison fair, I would like to keep all settings of the CARE network, but the model structure.

So far, I think I have figured it out until after the training, but then I am lost. I can't use model.predict() since that would require a CARE config (right?). Also, the other model requires that I strictly only predict images that have the right size (what functions did you use to make CARE applicable for all image sizes - I'd really like to understand that magic!).

So, my only workaround solution for the final image prediction would be to create patches myself, normalise them like in CARE, then predict the patches, stitch together - voila.

For that, I would need to understand, how the images are pre-processed in CARE before prediction. When you say "automatic percentile normalization" you mean normalization of the whole input image with percentiles that were calculated for that specific input image, right?

Thanks a lot,
Lisa

**Comment by maweigert** [comment]
> So, my only workaround solution for the final image prediction would be to create patches myself, normalise them like in CARE, then predict the patches, stitch together - voila.

Ok, now I understand. I would then simply normalize the image before stitching in the same way as for the other model and disable normalisation in `model.predict`  with `normalizer=None` :

```python
from csbdeep.utils import normalize 

img_normed = normalize(img, 1, 99.8)

model.predict(img_normed, axes = "YX", normalizer=None)
```

> (what functions did you use to make CARE applicable for all image sizes - I'd really like to understand that magic!).

Indeed, there is a lot of functionality inside `csbdeep` that tries to abstract these things away from the user and "just make it work" (e.g. image tiling, padding and cropping).  

> When you say "automatic percentile normalization" you mean normalization of the whole input image with percentiles that were calculated for that specific input image, right?

Yes, each image by default is normalized according to its respective percentiles.

**Comment by lizard24** [comment]
Perfect, I will try this!

Thanks a lot for your detailed answers, Martin!

**Comment by SJaffa** [comment]
Hi, we have come across an issue with the [`cbsdeep.utils.normalize`](https://github.com/CSBDeep/CSBDeep/blob/ad20e6d235efa205f175d63fb7c81b2c5e442922/csbdeep/utils/utils.py#L51) function as well and couldn't find this information in your documentation. Could you please add a description of the parameters to your documentation?

**Comment by maweigert** [comment]
Hi, 


> Hi, we have come across an issue with the [`cbsdeep.utils.normalize`](https://github.com/CSBDeep/CSBDeep/blob/ad20e6d235efa205f175d63fb7c81b2c5e442922/csbdeep/utils/utils.py#L51) function as well and couldn't find this information in your documentation.

what issue did you observe?

**Comment by SJaffa** [comment]
It was being used in another package with no explanation of the parameters. There is no problem with the code functionality, just that it is totally undocumented.

**Comment by maweigert** [comment]
> just that it is totally undocumented

Ah, now I understand (although I think the parameters are pretty self explanatory). We will add documentation when there is time.

**Comment by uschmidt83** [comment]
It's briefly mentioned [here](http://csbdeep.bioimagecomputing.com/doc/datagen.html?highlight=norm) and in section 2.2  of the [Supplementary Text and Figures](https://static-content.springer.com/esm/art%3A10.1038%2Fs41592-018-0216-7/MediaObjects/41592_2018_216_MOESM1_ESM.pdf).
